### PR TITLE
REF: implement Categorical._validate_setitem_value

### PR DESCRIPTION
**Pull request by jbrockmendel**
Moving towards sharing more methods

### CODE REVIEW

File: pandas/core/arrays/categorical.py [hunk 1]
```diff
@@ -1911,12 +1908,19 @@ def __setitem__(self, key, value):
...
 
-        # set by position
-        if isinstance(key, (int, np.integer)):
+        lindexer = self.categories.get_indexer(rvalue)
+        if isinstance(lindexer, np.ndarray) and lindexer.dtype.kind == "i":
+            lindexer = lindexer.astype(self._ndarray.dtype)
+
+        return lindexer
+
+    def _validate_setitem_key(self, key):
```
Comment (jreback) [inline pandas/core/arrays/categorical.py:1917]: can you type

File: pandas/core/arrays/datetimelike.py [hunk 2]
```diff
@@ -541,6 +541,15 @@ def __getitem__(self, key):
...
 
+        key = self._validate_getitem_key(key)
+        result = self._data[key]
+        if lib.is_scalar(result):
+            return self._box_func(result)
+
+        freq = self._get_getitem_freq(key)
+        return self._simple_new(result, dtype=self.dtype, freq=freq)
+
+    def _validate_getitem_key(self, key):
```
Comment (jreback) [inline pandas/core/arrays/datetimelike.py:552]: same

Comment (jreback) [reply pandas/core/arrays/datetimelike.py:552]: doc-string if easy

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
ok to type in later PRs, pls merge master

**Comment by jbrockmendel** [comment]
updated+green

**Review by jreback — APPROVED** [review]

**Pull request merged**